docs/source/cuda/external-memory.rst

```diff
@@ -0,0 +1,280 @@
+implementation follows:
+
+- Numba instantiates one instance of the EMM Plugin class per context. The
+  context that owns an EMM Plugin object is accessible through ``self.context``,
+  if required.
+- The EMM Plugin is transparent to any code that uses Numba - all its methods
+  are invoked by Numba, and never need to be called by code that uses Numba.
+- The allocation methods ``memalloc``, ``memhostalloc``, and ``mempin``, should
+  use the underlying library to allocate and/or pin device or host memory, and
+  construct an instance of a :ref:`memory pointer <memory-pointers>`
+  representing the memory to return back to Numba. These methods are always
+  called when the current CUDA context is the context that owns the EMM Plugin
+  instance.
+- The ``initialize`` method is called by Numba prior to the first use of the EMM
+  Plugin object for a context. This method should do anything required to
+  prepare the underlying library for allocations in the current context. This
+  method may be called multiple times, and must not invalidate previous state
+  when it is called.
+- The ``reset`` method is called when all allocations in the context are to be
+  cleaned up. It may be called even prior to ``initialize``, and an EMM Plugin
+  implementation needs to guard against this.
+- To support inter-GPU communication, the ``get_ipc_handle`` method should
+  provide an :class:`~numba.cuda.IpcHandle` for a given
+  :class:`~numba.cuda.MemoryPointer` instance. This method is part of the EMM
+  interface (rather than being handled within Numba) because the base address of
+  the allocation is only known by the underlying library. Closing an IPC handle
+  is handled internally within Numba.
+- It is optional provide memory info from the ``get_memory_info`` method, which
+  provides a count of the total and free memory on the device for the context.
+  It is preferrable to implement the method, but this may not be practical for
+  all allocators. If memory info is not provided, this method should raise a
+  :class:`RuntimeError`.
+- The ``defer_cleanup`` method should return a context manager that ensures that
+  expensive cleanup operations are avoided whilst it is active. The nuances of
+  this will vary between plugins, so the plugin documentation should include an
+  explanation of how deferring cleanup affects deallocations, and performance in
+  general.
+- The ``interface_version`` property is used to ensure that the plugin version
+  matches the interface provided by the version of Numba. At present, this
+  should always be 1.
+
+Full documentation for the base class follows:
+
+.. autoclass:: numba.cuda.BaseCUDAMemoryManager
+   :members: memalloc, memhostalloc, mempin, initialize, get_ipc_handle,
+             get_memory_info, reset, defer_cleanup, interface_version
+   :member-order: bysource
+
+
+.. _host-only-cuda-memory-manager:
+
+The Host-Only CUDA Memory Manager
+---------------------------------
+
```

Comment:
The memory manager is never host-only - that class is just to provide something to base an implementation of an EMM Plugin for libraries that don't want to manage host memory - e.g. the internal `NumbaCUDAMemoryManager` subclasses it, as does the RMM EMM Plugin (because RMM only does device memory at present).

There isn't a way to query what the memory manager is at the moment - ideally it should be transparent to the Numba user, so I'm finding it hard to see why you'd do something different when different memory managers are in use. I could add a `__repr__`, but if a `__repr__` should be provided, it's probably better to recommend that EMM Plugin authors implement `__repr__`, as `HostOnlyCUDAMemoryManager` is never instantiated directly (if you try the abstract base class checking will complain that `memalloc` etc. are not implemented).

Reply:
Perhaps that was a bad example, it was more about "is there likely to be a useful control flow pattern that is based on the type of memory manager", I now realise that this is unlikely the case and most probably something authors can just implement as they see fit.